packages/connect/src/connector/extension.ts

```diff
@@ -0,0 +1,133 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+
+import type {
+  ToApplication,
+  ToExtension,
+} from "@substrate/connect-extension-protocol"
+import type { Chain, JsonRpcCallback, SubstrateConnector } from "./types.js"
+import {
+  AlreadyDestroyedError,
+  CrashError,
+  JsonRpcDisabledError,
+} from "./errors.js"
+import { SupportedChains } from "../SupportedChains.js"
+
+type HeaderlessToExtensionGeneric<T extends ToExtension> = T extends {
+  origin: "substrate-connect-client"
+} & infer V
+  ? Omit<V, "chainId">
+  : unknown
+type HeaderlessToExtension = HeaderlessToExtensionGeneric<ToExtension>
+
+const listeners = new Map<string, (msg: ToApplication) => void>()
+window.addEventListener("message", ({ data }: MessageEvent<ToApplication>) => {
+  if (data?.origin !== "substrate-connect-extension") return
+  listeners.get(data.chainId)?.(data)
+})
+
+const listenToExtension = (
+  chainId: string,
+  callback: (msg: ToApplication) => void,
+): (() => void) => {
+  listeners.set(chainId, callback)
+  return () => {
+    listeners.delete(chainId)
+  }
+}
+
+function getRandomChainId(): string {
+  const arr = new BigUint64Array(2)
+  // It can only be used from the browser, so this is fine.
+  crypto.getRandomValues(arr)
+  const result = (arr[1] << BigInt(64)) | arr[0]
+  return result.toString(36)
+}
+
+export const getConnectorClient = (): SubstrateConnector => {
+  const chains = new Map<Chain, string>()
+
+  const internalAddChain = async (
+    isWellKnown: boolean,
+    chainSpecOrWellKnownName: string,
+    jsonRpcCallback?: JsonRpcCallback,
+    potentialRelayChainIds = [] as string[],
+  ): Promise<Chain> => {
+    const chainId = getRandomChainId()
+    const postToExtension = (msg: HeaderlessToExtension) => {
+      window.postMessage(
+        {
+          ...msg,
+          origin: "substrate-connect-client",
+          chainId,
+        },
+        "*",
+      )
+    }
+
+    await new Promise<void>((res, rej) => {
+      const stopListeningToExtension = listenToExtension(chainId, (msg) => {
+        stopListeningToExtension()
```

Comment:
I agree that this pattern when you register and it returns an unregistration function looks cool and all, but the goal is simplicity:

```suggestion
      listeners.set(chainId, (msg) => {
          listeners.delete(chainId);
```

Reply:
sure...